[test_all_gather_2d_fabric.py] | (Blackhole) e2e tests | blackhole-e2e-tests (QuietBox 2 (2xP300), bh_quietbox_2) / ccl nightly tests [bh_quietbox_2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] Blackhole CCL nightly tests hang with device timeout in test_all_gather_2d_fabric (system_memory_manager.cpp:780)
<!-- AUTO-TRIAGE-SUMMARY-END -->
**Workflow:** (Blackhole) e2e tests
**Job:** blackhole-e2e-tests (QuietBox 2 (2xP300), bh_quietbox_2) / ccl nightly tests [bh_quietbox_2]

### Regression timeline
- Last passing commit: `c4e3dfe30095` (2026-07-19T08:43:37Z) — https://github.com/tenstorrent/tt-metal/actions/runs/29680310452
- First failing commit: `8c3c1f3c787e` (2026-07-27T09:01:35Z) — https://github.com/tenstorrent/tt-metal/actions/runs/30252180262/job/90038592682
- Most recent failing commit: `2330af73f529` (2026-07-27T16:40:00Z) — https://github.com/tenstorrent/tt-metal/actions/runs/30285732797/job/90114394493

### Failing job URLs (last 2 runs)
- https://github.com/tenstorrent/tt-metal/actions/runs/30285732797/job/90114394493
- https://github.com/tenstorrent/tt-metal/actions/runs/30252180262/job/90038592682

### Failing test path(s)
- tests/ttnn/unit_tests/operations/ccl/blackhole_CI/box/nightly/test_all_gather_2d_fabric.py::test_all_gather_2d_fabric_linear (first failure in both runs)
- tests/ttnn/unit_tests/operations/ccl/blackhole_CI/box/nightly/test_all_gather_2d_fabric_turning.py::test_all_gather_2d_fabric_turning (also fails identically in both runs)
- Additional cascading failures in run 1 only (test_minimal_reduce_scatter_async_bh.py tests) — device becomes unrecoverable after the first hang, so subsequent tests in the same job also fail with the same throw.

### Error excerpt (terminal failure from logs)
```
2026-07-27 17:56:56.486 | critical | Always | TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable (assert.hpp:104)
##[error]test_all_gather_2d_fabric_linear[silicon_arch_name=blackhole-silicon_arch_blackhole=True-cluster_axis=0-device_params={'fabric_config': FabricConfig.FABRIC_2D, 'trace_region_size': 90112}-num_iters=3-enable_trace=False-...]

RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception
```
```
# Run 2 (identical):
2026-07-27 09:44:25.910 | critical | Always | TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable (assert.hpp:104)
FAILED tests/ttnn/unit_tests/operations/ccl/blackhole_CI/box/nightly/test_all_gather_2d_fabric.py::test_all_gather_2d_fabric_linear[...] - RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception
```

### Reproduction steps
1. On a Blackhole QuietBox (2xP300) host, run:
```
pytest tests/ttnn/unit_tests/operations/ccl/blackhole_CI/box/nightly
```
2. Observe `test_all_gather_2d_fabric_linear` (and `test_all_gather_2d_fabric_turning`) hang and throw a device timeout during `ttnn.synchronize_device(mesh_device, sub_device_ids=sub_device_stall_group)`.

### Notes
This is a device hang (timeout), not an assertion/logic failure — the device becomes "unrecoverable" after the timeout, which explains the additional cascading failures seen later in run 1 (subsequent reduce_scatter tests fail with the same TT_THROW since the device was never reset).

**Determinism check:**
- Run 1 terminal error (first failing test): `TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable` → `RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception` on `test_all_gather_2d_fabric_linear`
- Run 2 terminal error (first failing test): identical `TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable` → `RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception` on the same `test_all_gather_2d_fabric_linear`
- Both runs also fail `test_all_gather_2d_fabric_turning` with the identical error immediately after.
- Last passing and first/most-recent failing runs reference distinct commits/run URLs.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (Blackhole) e2e tests`
`Auto-triage-job-name: blackhole-e2e-tests (QuietBox 2 (2xP300), bh_quietbox_2) / ccl nightly tests [bh_quietbox_2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/29846161351
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1784651333004499?thread_ts=1784649253.026159&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_unicast_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_unicast_factory.cpp
@@ -14,7 +14,7 @@ namespace ttnn::operations::ccl {
 using namespace ::ttnn::ccl;
 
 ////////////////////////////////////////////////////////////////
-// Store-and-forward AllGather (Fabric_1D line/ring only)
+// Store-and-forward AllGather (line/ring over a single mesh axis; for Fabric 1D and 2D)
 //
 // Every device relays stripes to its neighbor one hop at a time; a shard reaches far devices by being
 // re-forwarded at each hop. Forward and backward directions run on separate cores. Per direction: the reader
@@ -539,23 +539,28 @@ AllGatherUnicastFactory::cached_program_t AllGatherUnicastFactory::create_at(
                 };
                 tt::tt_metal::SetRuntimeArgs(program, reader_kernel_id, {core}, reader_rt_args);
 
+                // route for Fabric_2D
+                const auto route_node =
+                    neighbor.has_value() ? mesh_device->get_fabric_node_id(*neighbor) : sender_fabric_node_id;
                 std::vector<uint32_t> writer_rt_args = {
-                    output_addr,               // output tensor address
-                    device_idx,                // this device's index (initial stripe)
-                    stripe_step,               // stripe index step per iteration
-                    num_iters,                 // iterations this direction runs
-                    local_output_start,        // this worker's slice start (chunks)
-                    num_worker_output_chunks,  // this worker's slice length (chunks)
-                    final_start,               // last-iteration slice start (even-ring split)
-                    final_count,               // last-iteration slice length (even-ring split)
-                    do_local_write ? 1u : 0u,  // write local data into local output on iteration 0
-                    barrier_sem.address(),     // barrier_sem L1 address
-                    data_valid_sem.address(),  // data_valid_sem L1 address
-                    (uint32_t)partner_core.x,  // barrier_sem target (neighbor partner core x)
-                    (uint32_t)partner_core.y,  // barrier_sem target (neighbor partner core y)
-                    (uint32_t)mirror_core.x,   // data_valid_sem target (neighbor mirror core x)
-                    (uint32_t)mirror_core.y,   // data_valid_sem target (neighbor mirror core y)
-                    num_granular,              // leading sends the downstream relays
+                    output_addr,                      // output tensor address
+                    device_idx,                       // this device's index (initial stripe)
+                    stripe_step,                      // stripe index step per iteration
+                    num_iters,                        // iterations this direction runs
+                    local_output_start,               // this worker's slice start (chunks)
+                    num_worker_output_chunks,         // this worker's slice length (chunks)
+                    final_start,                      // last-iteration slice start (even-ring split)
+                    final_count,                      // last-iteration slice length (even-ring split)
+                    do_local_write ? 1u : 0u,         // write local data into local output on iteration 0
+                    barrier_sem.address(),            // barrier_sem L1 address
+                    data_valid_sem.address(),         // data_valid_sem L1 address
+                    (uint32_t)partner_core.x,         // barrier_sem target (neighbor partner core x)
+                    (uint32_t)partner_core.y,         // barrier_sem target (neighbor partner core y)
+                    (uint32_t)mirror_core.x,          // data_valid_sem target (neighbor mirror core x)
+                    (uint32_t)mirror_core.y,          // data_valid_sem target (neighbor mirror core y)
+                    num_granular,                     // leading sends the downstream relays
+                    (uint32_t)route_node.chip_id,     // neighbor chip id (packet header 2D route)
+                    (uint32_t)(*route_node.mesh_id),  // neighbor mesh id (packet header 2D route)
                 };
                 TT_FATAL(num_iters == 0 || neighbor.has_value(), "an active direction must have a neighbor");
                 if (num_iters > 0) {

--- a/ttnn/cpp/ttnn/operations/ccl/all_gather/device/kernels/unicast_common.hpp
+++ b/ttnn/cpp/ttnn/operations/ccl/all_gather/device/kernels/unicast_common.hpp
@@ -9,9 +9,11 @@
 
 #include <array>
 #include <cstdint>
+#include <type_traits>
 #include <utility>
 
-// Store-and-forward AllGather is Fabric_1D only: every fabric send is a single 1-hop unicast to the neighbor.
+// Store-and-forward AllGather: every fabric send is a single 1-hop unicast to the neighbor.
+// Runs on any effectively-1D topology (both Fabric 1D and 2D).
 namespace fabric_api = tt::tt_fabric::linear::experimental;
 
 ////////////////////////////////////////////////////////////////
@@ -93,20 +95,22 @@ private:
 };
 
 // Unicasts pages one hop to the single neighbor. Handles packetization (pack several pages into one
-// scatter-write packet when they fit, else split a big page across packets).
+// scatter-write packet when they fit, else split a big page across packets), and the semaphore increments
+// that announce them.
 //
 // Templated on the sender type (SenderT*) so the same writer drives either a direct WorkerToFabricEdmSender
 // (one worker per direction) or a WorkerToFabricMuxSender (workers sharing a fabric mux). The send calls are
-// base sender-pointer overloads that accept either; for 1D-linear all routing is to_chip_unicast(1) set via
-// set_state, so no route-manager is needed. FabricWriter owns its two packet headers directly.
+// base sender-pointer overloads that accept either, so no route-manager is needed -- which is also why this
+// class routes its own headers.
 template <uint32_t page_size, uint32_t packet_size, typename SenderT>
 class FabricWriter {
 public:
-    FabricWriter(const Noc& noc, SenderT* sender) :
+    FabricWriter(const Noc& noc, SenderT* sender, uint16_t neighbor_chip_id, uint16_t neighbor_mesh_id) :
         noc{noc},
         sender{sender},
         scatter_packet_header{PacketHeaderPool::allocate_header(1)},
         unicast_packet_header{PacketHeaderPool::allocate_header(1)},
+        sem_packet_header{PacketHeaderPool::allocate_header(1)},
         scatter_header({}, {}),
         chunk_count{0} {
         std::array<uint64_t, max_pages_per_packet> dummy_addrs{};  // init to 0s
@@ -121,10 +125,35 @@ public:
 
         fabric_api::fabric_unicast_noc_unicast_write_set_state<UnicastWriteUpdateMask::None>(
             unicast_packet_header, num_hops);
+
+        // One atomic-inc header for the "alive" barrier inc + data_valid signals; Flush orders it after the
+        // payload it announces.
+        fabric_api::fabric_unicast_noc_unicast_atomic_inc_set_state<
+            UnicastAtomicIncUpdateMask::Val | UnicastAtomicIncUpdateMask::Flush>(
+            sem_packet_header, num_hops, tt::tt_fabric::NocUnicastAtomicIncCommandHeader{0u, 1u});
+
+        // For Fabric_2D, set_state() sets routes only in its RoutingPlaneConnectionManager overloads, so set
+        // them here. Keyed on the header type since the FABRIC_2D define is absent on the mux path.
+        if constexpr (std::is_base_of_v<tt::tt_fabric::HybridMeshPacketHeader, PACKET_HEADER_TYPE>) {
+            using MeshHeader = volatile tt::tt_fabric::HybridMeshPacketHeader*;
+            fabric_set_unicast_route(
+                reinterpret_cast<MeshHeader>(scatter_packet_header), neighbor_chip_id, neighbor_mesh_id);
+            fabric_set_unicast_route(
+                reinterpret_cast<MeshHeader>(unicast_packet_header), neighbor_chip_id, neighbor_mesh_id);
+            fabric_set_unicast_route(
+                reinterpret_cast<MeshHeader>(sem_packet_header), neighbor_chip_id, neighbor_mesh_id);
+        }
     }
 
     ~FabricWriter() {
         ASSERT(chunk_count == 0);  // outstanding chunks! flush() not called correctly
+    }
+
+    // Increment a semaphore on the neighbor.
+    void atomic_inc(uint64_t addr, uint32_t val) {
+        fabric_api::fabric_unicast_noc_unicast_atomic_inc_with_state<
+            UnicastAtomicIncUpdateMask::DstAddr | UnicastAtomicIncUpdateMask::Val>(
+            sender, sem_packet_header, tt::tt_fabric::NocUnicastAtomicIncCommandHeader{addr, val});
     }
 
     void async_write(uint32_t l1_addr, uint64_t remote_noc_addr) {
@@ -211,6 +240,7 @@ private:
     SenderT* sender;  // direct or mux sender
     volatile tt_l1_ptr PACKET_HEADER_TYPE* scatter_packet_header;
     volatile tt_l1_ptr PACKET_HEADER_TYPE* unicast_packet_header;
+    volatile tt_l1_ptr PACKET_HEADER_TYPE* sem_packet_header;
     NocUnicastScatterCommandHeader scatter_header;
     uint8_t chunk_count;     // accumulated chunks not yet sent in a packet
     uint32_t start_l1_addr;  // start address of the accumulated contiguous chunks

--- a/ttnn/cpp/ttnn/operations/ccl/all_gather/device/kernels/unicast_writer.cpp
+++ b/ttnn/cpp/ttnn/operations/ccl/all_gather/device/kernels/unicast_writer.cpp
@@ -71,6 +71,8 @@ void kernel_main() {
     const uint8_t data_valid_sem_noc_x = get_arg_val<uint32_t>(arg_idx++);  // mirror core (data_valid_sem target)
     const uint8_t data_valid_sem_noc_y = get_arg_val<uint32_t>(arg_idx++);
     const uint32_t num_granular_sends = get_arg_val<uint32_t>(arg_idx++);  // leading sends the downstream relays
+    const uint16_t neighbor_chip_id = get_arg_val<uint32_t>(arg_idx++);
+    const uint16_t neighbor_mesh_id = get_arg_val<uint32_t>(arg_idx++);
     [[maybe_unused]] size_t arg_for_fab = arg_idx;  // fabric connection args start here (non-mux path)
 
     // A direction with no neighbor (a line endpoint) relays nothing; no fabric/mux connection was appended.
@@ -139,29 +141,17 @@ void kernel_main() {
     SenderT* sender = &fabric_connection.get(0).sender;
 #endif
 
-    FabricWriter<output_chunk_size, packet_size, SenderT> fabric(noc, sender);
-
-    // One 1-hop atomic-inc header for both the "alive" barrier inc and the data_valid signals; destination and
-    // value (chunks) are set per send. Flush keeps a data_valid inc ordered after the payload it announces.
-    auto sem_packet_header = PacketHeaderPool::allocate_header(1);
-    fabric_api::fabric_unicast_noc_unicast_atomic_inc_set_state<
-        UnicastAtomicIncUpdateMask::Val | UnicastAtomicIncUpdateMask::Flush>(
-        sem_packet_header, /*num_hops=*/1, tt::tt_fabric::NocUnicastAtomicIncCommandHeader{0u, 1u});
-    auto atomic_inc = [&](uint64_t addr, uint32_t val) {
-        fabric_api::fabric_unicast_noc_unicast_atomic_inc_with_state<
-            UnicastAtomicIncUpdateMask::DstAddr | UnicastAtomicIncUpdateMask::Val>(
-            sender, sem_packet_header, tt::tt_fabric::NocUnicastAtomicIncCommandHeader{addr, val});
-    };
+    FabricWriter<output_chunk_size, packet_size, SenderT> fabric(noc, sender, neighbor_chip_id, neighbor_mesh_id);
 
     // Init handshake (send only): tell the neighbor's opposite-direction reader we're alive, so it lets its
     // paired writer start writing into our output. Our own reader does the matching wait.
     if constexpr (do_init_barrier) {
-        atomic_inc(safe_get_noc_addr(barrier_sem_noc_x, barrier_sem_noc_y, barrier_sem, 0), 1);
+        fabric.atomic_inc(safe_get_noc_addr(barrier_sem_noc_x, barrier_sem_noc_y, barrier_sem, 0), 1);
     }
 
     const uint64_t downstream_data_valid_addr =
         safe_get_noc_addr(data_valid_sem_noc_x, data_valid_sem_noc_y, data_valid_sem, 0);
-    auto signal = [&](uint32_t chunks) { atomic_inc(downstream_data_valid_addr, chunks); };
+    auto signal = [&](uint32_t chunks) { fabric.atomic_inc(downstream_data_valid_addr, chunks); };
 
     ///////////////////////////////////////////////////
     // MAIN

